WireDomain/Sources/WireDomain/Repositories/FeatureConfigRepository.swift

```diff
@@ -0,0 +1,352 @@
+//
+// Wire
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+import WireAPI
+import WireDataModel
+
+/// Facilitate access to feature configs related domain objects.
+
+protocol FeatureConfigRepositoryProtocol {
+
+    /// Pull feature configs from the server and store them locally.
+    ///
+    /// The `AsyncStream` produces a `FeatureState` value right after the feature is stored locally.
+    /// so actions can be triggered right away at a lower level callsite (e.g an interactor).
+    /// This will allow a user to be notified immediately even if the stream fails afterwards.
+    ///
+    /// - Returns: An async stream of `FeatureState`.
+
+    func pullFeatureConfigs() -> AsyncThrowingStream<FeatureState, Error>
+
+    /// Fetch a feature config locally.
+    ///
+    /// - Parameter name: The feature name to fetch the config for.
+    /// - Parameter type: The type of config to retrieve.
+    /// - Returns: A `LocalFeature` object with a status and a config (if any).
+
+    func fetchFeatureConfig<T: Decodable>(withName name: Feature.Name, type: T.Type) async throws -> LocalFeature<T>
+
+    /// Fetches a flag indicating whether the user should be notified of a given feature.
+    /// - Parameter name: The feature name.
+    /// - Returns: `true` if user should be notified.
+
+    func fetchNeedsToNotifyUser(forFeatureName name: Feature.Name) async throws -> Bool
+
+    /// Stores a flag indicating whether the user should be notified of a given feature.
+    /// - Parameter notifyUser: Whether the user should be notified for a given feature.
+    /// - Parameter name: The name of the feature to set the flag for.
+
+    func storeNeedsToNotifyUser(_ notifyUser: Bool, forFeatureName name: Feature.Name) async throws
+
+}
+
+final class FeatureConfigRepository: FeatureConfigRepositoryProtocol {
+
+    // MARK: - Properties
+
+    private let featureConfigsAPI: any FeatureConfigsAPI
+    private let context: NSManagedObjectContext
+
+    // MARK: - Object lifecycle
+
+    init(
+        featureConfigsAPI: any FeatureConfigsAPI,
+        context: NSManagedObjectContext
+    ) {
+        self.featureConfigsAPI = featureConfigsAPI
+        self.context = context
+    }
+
+    // MARK: - Public
+
+    func pullFeatureConfigs() -> AsyncThrowingStream<FeatureState, Error> {
```

Comment:
No network request between each, it was mostly designed as a callback mechanism to let the caller know right away when a feature has been processed and take action consequently (even if an error is thrown later as you mentioned). I wanted to stick to how it is currently designed in legacy codebase (as soon as a feature is stored it updates some UI related stuff) and process to the next one.

Reply:
Oh I see. @johnxnguyen maybe you have an opinion here? Does the legacy behavior make sense to you?

Reply:
Good idea, it makes sense. However, I would not introduce any Combine code since we can still use async / await code to leverage this behavior while remaining consistent in our approach. I will give it a try

Reply:
@johnxnguyen my bad, I thought AsyncStream could behave quite similarly as a publisher but for our purpose that is we want to keep our stream 'alive' and receive values over time, a publisher is more suitable. So did the fix you mentioned  🙏

Reply:
We could also have a long lived task with an infinite stream, whatever makes most sense